src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,291 @@
+            {
+                context.RegisterRefactoring(new MyCodeAction(
+                    FeaturesResources.Introduce_using_statement,
+                    cancellationToken => IntroduceUsingStatementAsync(document, span, cancellationToken)));
+            }
+        }
+
+        private static async Task<(TLocalDeclarationSyntax, ILocalSymbol)> FindDisposableLocalDeclaration(Document document, TextSpan selection, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var declarationSyntax =
+                root.FindNode(selection)?.GetAncestor<TLocalDeclarationSyntax>()
+                ?? root.FindTokenOnLeftOfPosition(selection.End).GetAncestor<TLocalDeclarationSyntax>();
+
+            if (declarationSyntax is null)
+            {
+                return default;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var operation = semanticModel.GetOperation(declarationSyntax, cancellationToken) as IVariableDeclarationGroupOperation;
+            if (operation?.Declarations.Length != 1)
+            {
+                return default;
+            }
+
+            var localDeclaration = operation.Declarations[0];
+            if (localDeclaration.Declarators.Length != 1)
+            {
+                return default;
+            }
+
+            var declarator = localDeclaration.Declarators[0];
+            if (declarator.Initializer is null)
+            {
+                return default;
+            }
+
+            var localType = declarator.Symbol?.Type;
+            if (localType is null)
+            {
+                return default;
+            }
+
+            if (declarator.Initializer.Value is null ||
+                declarator.Initializer.Value.Kind == OperationKind.Invalid)
+            {
+                return default;
+            }
+
+            var isValidSelection = await CodeRefactoringHelpers.RefactoringSelectionIsValidAsync(
+                document,
+                selection,
+                node: declarationSyntax,
+                holes: ImmutableArray.Create(declarator.Initializer.Value.Syntax),
+                cancellationToken).ConfigureAwait(false);
+
+            if (!isValidSelection)
+            {
+                return default;
+            }
+
+            if (!IsLegalUsingStatementType(semanticModel.Compilation, localType))
+            {
+                return default;
+            }
+
+            return (declarationSyntax, declarator.Symbol);
+        }
+
+        private static bool IsLegalUsingStatementType(Compilation compilation, ITypeSymbol type)
+        {
+            var disposableType = compilation.GetSpecialType(SpecialType.System_IDisposable);
+            if (disposableType is null)
+            {
+                return default;
+            }
```

Comment:
consider doing this check up front after the syntactic checks, but before the IOperation checks.

Reply:
It uses IOperation in order to find the syntax for the hole, so it has to be last or else be specialized for every language.

Reply:
didn't you do this check already?